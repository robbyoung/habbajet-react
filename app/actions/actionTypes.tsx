export enum ActionType {
    ADD_HABBAJET = 'Add Habbajet',
    ADD_HABIT_RESULT = 'Add Habit Result',
    UPDATE_BUDGET = 'Update Budget',
    RESET_HABBAJET = 'Reset Habbajet',
    LOAD_STATE = 'Load State',
    SELECT_HABBAJET = 'Select Habbajet',
    UPDATE_EDITOR_FIELD = 'Update Editor Field',
    VALIDATE_EDITOR = 'Validate Editor',
    CLEAR_EDITOR = 'Clear Editor',
    ADD_PURCHASE = 'Add Purchase',
    UPDATE_PURCHASE_EDITOR = 'Update Purchase Editor',
    CLEAR_PURCHASE_EDITOR = 'Clear Purchase Editor',
    VALIDATE_PURCHASE_EDITOR = 'Validate Purchase Editor',
    SET_HABBAJET_TO_EDIT = 'Set Habbajet To Edit',
    DELETE_HABBAJET = 'Delete Habbajet',
    ADD_TAG = 'Add Tag',
    SET_PURCHASE_TO_EDIT = 'Set Purchase To Edit',
    SET_TAG_TO_EDIT = 'Set Tag To Edit',
    UPDATE_TAG_EDITOR = 'Update Tag Editor',
    CLEAR_TAG_EDITOR = 'Clear Tag Editor',
    VALIDATE_TAG_EDITOR = 'Validate Tag Editor',
    DELETE_TAG = 'Delete Tag',
}
