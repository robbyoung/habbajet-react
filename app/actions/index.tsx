export * from './actionTypes';
export * from './addHabbajet';
export * from './addHabitResult';
export * from './updateBudget';
export * from './resetHabbajet';
export * from './loadState';
export * from './selectHabbajet';
export * from './updateEditorField';
export * from './validateEditor';
export * from './clearEditor';
