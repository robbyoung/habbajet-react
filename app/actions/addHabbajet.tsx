import moment from 'moment';
import {Action} from 'redux';
import {ActionType} from './actionTypes';
import {Habbajet} from '../state';

export interface AddHabbajetAction extends Action {
    type: ActionType.ADD_HABBAJET;
    newHabbajet: Habbajet;
}

export function addHabbajetAction(
    name: string,
    value: number,
    modifier: number,
    slack: number,
    color: string,
): AddHabbajetAction {
    const monday = moment()
        .startOf('isoWeek')
        .toISOString();
    return {
        type: ActionType.ADD_HABBAJET,
        newHabbajet: {
            name,
            maxValue: value,
            currentValue: value / Math.pow(modifier, 7),
            modifier,
            results: [],
            color,
            date: monday,
            toClaim: false,
            selected: false,
            currentStreak: 0,
            bestStreak: 0,
            totalSlack: slack,
            remainingSlack: slack,
        },
    };
}

export function addHabbajet(
    state: Habbajet[],
    action: AddHabbajetAction,
): Habbajet[] {
    const replaceIndex = state.findIndex(
        habbajet => habbajet.name === action.newHabbajet.name,
    );

    if (replaceIndex === -1) {
        return [...state, action.newHabbajet];
    }

    const newState = [...state];
    const edited = action.newHabbajet;
    const original = state[replaceIndex];

    newState[replaceIndex] = {
        ...original,
        name: edited.name,
        maxValue: edited.maxValue,
        modifier: edited.modifier,
        totalSlack: edited.totalSlack,
        color: edited.color,
        selected: true,
    };

    return newState;
}
