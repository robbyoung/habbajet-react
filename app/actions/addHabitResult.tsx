import moment from 'moment';
import {Action} from 'redux';
import {ActionType} from './actionTypes';
import {Habbajet, HabitResult} from '../state';

export interface AddHabitResultAction extends Action {
    type: ActionType.ADD_HABIT_RESULT;
    habbajetName: string;
    result: boolean;
}

export function addHabitResultAction(
    habbajetName: string,
    result: boolean,
): AddHabitResultAction {
    return {
        type: ActionType.ADD_HABIT_RESULT,
        habbajetName,
        result,
    };
}

export function addHabitResult(
    state: Habbajet[],
    action: AddHabitResultAction,
): Habbajet[] {
    const index = state.findIndex(
        habbajet => habbajet.name === action.habbajetName,
    );
    if (index === -1) {
        return state;
    }

    const newState = [...state];
    const edited: Habbajet = {...newState[index]};
    const date = moment(edited.date);

    date.add(1, 'day');
    if (date.day() === 1) {
        edited.toClaim = true;
    }
    edited.date = date.toISOString();

    let newResult = action.result ? HabitResult.Success : HabitResult.Failure;
    if (edited.remainingSlack > 0 && !action.result) {
        newResult = HabitResult.SlackSuccess;
    }
    edited.results = [...edited.results, newResult];

    const isSuccess = action.result || edited.remainingSlack > 0;
    if (isSuccess) {
        edited.currentStreak++;
        if (edited.currentStreak > edited.bestStreak) {
            edited.bestStreak = edited.currentStreak;
        }
        if (!action.result) {
            edited.remainingSlack--;
        }
    } else {
        edited.currentValue /= edited.modifier;
        edited.currentStreak = 0;
    }

    newState[index] = edited;
    return newState;
}
