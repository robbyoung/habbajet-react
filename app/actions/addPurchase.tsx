import moment from 'moment';
import {Action} from 'redux';
import {ActionType} from './actionTypes';
import {Purchase} from '../state';

export interface AddPurchaseAction extends Action {
    type: ActionType.ADD_PURCHASE;
    newPurchase: Purchase;
}

export function addPurchaseAction(
    name: string,
    cost: number,
    tagId: string,
): AddPurchaseAction {
    return {
        type: ActionType.ADD_PURCHASE,
        newPurchase: {
            name,
            cost,
            tagId,
            date: moment().toISOString(),
        },
    };
}

export function addPurchase(
    state: Purchase[],
    action: AddPurchaseAction,
): Purchase[] {
    return [action.newPurchase, ...state];
}
