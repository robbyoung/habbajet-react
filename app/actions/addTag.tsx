import {Action} from 'redux';
import {ActionType} from './actionTypes';
import {Tag} from '../state';
import uuid from 'react-native-uuid';

export interface AddTagAction extends Action {
    type: ActionType.ADD_TAG;
    newTag: Tag;
}

export function addTagAction(name: string, color: string): AddTagAction {
    return {
        type: ActionType.ADD_TAG,
        newTag: {
            name,
            color,
            id: uuid.v4(),
        },
    };
}

export function addTag(state: Tag[], action: AddTagAction): Tag[] {
    return [...state, action.newTag];
}
