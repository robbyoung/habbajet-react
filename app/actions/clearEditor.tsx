import {Action} from 'redux';
import {ActionType} from './actionTypes';
import {HabbajetEditor} from '../state';
import {DEFAULT_EDITOR_STATE} from '../state/defaults';

export interface ClearEditorAction extends Action {
    type: ActionType.CLEAR_EDITOR;
}

export function clearEditorAction(): ClearEditorAction {
    return {
        type: ActionType.CLEAR_EDITOR,
    };
}

export function clearEditor(
    _state: HabbajetEditor,
    _action: ClearEditorAction,
): HabbajetEditor {
    return DEFAULT_EDITOR_STATE;
}
