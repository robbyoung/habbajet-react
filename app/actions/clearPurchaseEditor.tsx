import {Action} from 'redux';
import {ActionType} from './actionTypes';
import {PurchaseEditor} from '../state';
import {DEFAULT_PURCHASE_EDITOR_STATE} from '../state/defaults';

export interface ClearPurchaseEditorAction extends Action {
    type: ActionType.CLEAR_PURCHASE_EDITOR;
}

export function clearPurchaseEditorAction(): ClearPurchaseEditorAction {
    return {
        type: ActionType.CLEAR_PURCHASE_EDITOR,
    };
}

export function clearPurchaseEditor(
    _state: PurchaseEditor,
    _action: ClearPurchaseEditorAction,
): PurchaseEditor {
    return DEFAULT_PURCHASE_EDITOR_STATE;
}
