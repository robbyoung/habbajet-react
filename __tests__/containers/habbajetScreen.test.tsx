import React from 'react';
import renderer from 'react-test-renderer';
import HabbajetScreen from '../../app/containers/habbajetScreen';
import store from '../../app/store';
import { addHabbajetAction } from '../../app/actions';
import { Provider } from 'react-redux';

jest.mock('@fortawesome/react-native-fontawesome', () => ({
    FontAwesomeIcon: '',
}));

describe('Habbajet Screen Component', () => {
    it('will map state to a habbajet display', () => {
        store.dispatch(addHabbajetAction('Test Habbajet', 100, '#8066C2'));

        const component = renderer.create(
            <Provider store={store}>
                <HabbajetScreen/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});
