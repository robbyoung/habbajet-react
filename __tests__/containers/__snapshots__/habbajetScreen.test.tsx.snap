// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Habbajet Screen Component will map state to a habbajet display 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": "100%",
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "margin": 30,
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <ARTSurfaceView
          style={
            Array [
              undefined,
              Object {
                "height": 690,
                "width": 690,
              },
            ]
          }
        >
          <ARTGroup
            opacity={1}
            transform={
              Array [
                6.123233995736766e-17,
                -1,
                1,
                6.123233995736766e-17,
                0,
                690,
              ]
            }
          >
            <ARTShape
              d={
                Array [
                  0,
                  517.5,
                  345,
                  4,
                  345,
                  345,
                  172.5,
                  0,
                  6.283185307179586,
                  1,
                ]
              }
              fill={null}
              opacity={1}
              stroke="#ffffff"
              strokeCap={1}
              strokeDash={null}
              strokeJoin={1}
              strokeWidth={345}
              transform={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
            />
            <ARTShape
              d={
                Array [
                  0,
                  517.394917660794,
                  351.02016318118143,
                  4,
                  345,
                  345,
                  172.5,
                  0.03490658503988659,
                  0.8975979010256553,
                  1,
                ]
              }
              fill={null}
              opacity={1}
              stroke="#f5f5f5"
              strokeCap={0}
              strokeDash={null}
              strokeJoin={1}
              strokeWidth={345}
              transform={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
            />
            <ARTShape
              d={
                Array [
                  0,
                  447.7797199491386,
                  483.53728439368456,
                  4,
                  345,
                  345,
                  172.5,
                  0.9325044860655419,
                  1.7951958020513106,
                  1,
                ]
              }
              fill={null}
              opacity={1}
              stroke="#f5f5f5"
              strokeCap={0}
              strokeDash={null}
              strokeJoin={1}
              strokeWidth={345}
              transform={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
            />
            <ARTShape
              d={
                Array [
                  0,
                  300.769296791575,
                  511.7330048121494,
                  4,
                  345,
                  345,
                  172.5,
                  1.8301023870911974,
                  2.692793703076966,
                  1,
                ]
              }
              fill={null}
              opacity={1}
              stroke="#f5f5f5"
              strokeCap={0}
              strokeDash={null}
              strokeJoin={1}
              strokeWidth={345}
              transform={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
            />
            <ARTShape
              d={
                Array [
                  0,
                  187.06549529187478,
                  414.375371873592,
                  4,
                  345,
                  345,
                  172.5,
                  2.7277002881168526,
                  3.590391604102621,
                  1,
                ]
              }
              fill={null}
              opacity={1}
              stroke="#f5f5f5"
              strokeCap={0}
              strokeDash={null}
              strokeJoin={1}
              strokeWidth={345}
              transform={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
            />
            <ARTShape
              d={
                Array [
                  0,
                  192.28959711417252,
                  264.77666891453424,
                  4,
                  345,
                  345,
                  172.5,
                  3.625298189142508,
                  4.487989505128277,
                  1,
                ]
              }
              fill={null}
              opacity={1}
              stroke="#f5f5f5"
              strokeCap={0}
              strokeDash={null}
              strokeJoin={1}
              strokeWidth={345}
              transform={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
            />
            <ARTShape
              d={
                Array [
                  0,
                  312.50774703402135,
                  175.5877705205588,
                  4,
                  345,
                  345,
                  172.5,
                  4.522896090168163,
                  5.385587406153932,
                  1,
                ]
              }
              fill={null}
              opacity={1}
              stroke="#f5f5f5"
              strokeCap={0}
              strokeDash={null}
              strokeJoin={1}
              strokeWidth={345}
              transform={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
            />
            <ARTShape
              d={
                Array [
                  0,
                  457.19322615842384,
                  213.96973630429966,
                  4,
                  345,
                  345,
                  172.5,
                  5.420493991193819,
                  6.283185307179588,
                  1,
                ]
              }
              fill={null}
              opacity={1}
              stroke="#f5f5f5"
              strokeCap={0}
              strokeDash={null}
              strokeJoin={1}
              strokeWidth={345}
              transform={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
            />
            <ARTGroup
              opacity={1}
              transform={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
            />
          </ARTGroup>
        </ARTSurfaceView>
      </View>
      <View
        style={
          Object {
            "marginBottom": 15,
            "marginHorizontal": "10%",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Abel",
                  "fontSize": 20,
                  "margin": 0,
                  "padding": 0,
                },
                Object {
                  "color": "#E04747",
                },
              ]
            }
          >
            Expected Output
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Abel",
                  "marginTop": -5,
                  "padding": 0,
                },
                Object {
                  "fontSize": 50,
                },
              ]
            }
          >
            $100.00
          </Text>
        </View>
        <View
          style={
            Object {
              "marginBottom": 10,
              "marginTop": 10,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Abel",
                  "fontSize": 20,
                  "margin": 0,
                  "padding": 0,
                },
                Object {
                  "color": "#E04747",
                },
              ]
            }
          >
            Streaks
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontFamily": "Abel",
                  "fontSize": 20,
                  "margin": 0,
                  "padding": 0,
                  "width": 35,
                }
              }
            >
              Now
            </Text>
            <View
              style={
                Array [
                  Object {
                    "height": 20,
                    "marginRight": 5,
                  },
                  Object {
                    "backgroundColor": "#E04747",
                    "width": 2,
                  },
                ]
              }
            />
            <Text
              style={
                Object {
                  "fontFamily": "Abel",
                  "fontSize": 20,
                  "margin": 0,
                  "padding": 0,
                }
              }
            >
              0
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontFamily": "Abel",
                  "fontSize": 20,
                  "margin": 0,
                  "padding": 0,
                  "width": 35,
                }
              }
            >
              Best
            </Text>
            <View
              style={
                Array [
                  Object {
                    "height": 20,
                    "marginRight": 5,
                  },
                  Object {
                    "backgroundColor": "#E04747",
                    "width": 450,
                  },
                ]
              }
            />
            <Text
              style={
                Object {
                  "fontFamily": "Abel",
                  "fontSize": 20,
                  "margin": 0,
                  "padding": 0,
                }
              }
            >
              5
            </Text>
          </View>
        </View>
        <View />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "flexDirection": "column",
          "padding": 10,
        },
        Object {
          "backgroundColor": "#E04747",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "#ffffff",
            "fontFamily": "Abel",
            "fontSize": 30,
            "textAlign": "center",
            "width": "100%",
          },
        ]
      }
    >
      Monday, March 23rd
    </Text>
    <View
      style={
        Array [
          Object {
            "alignContent": "center",
            "flexDirection": "row",
            "justifyContent": "space-around",
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={false}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="button-success"
      >
        <
          color="#ffffff"
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f00c",
                "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z",
              ],
              "iconName": "check",
              "prefix": "fas",
            }
          }
          size={100}
        />
      </View>
      <View
        accessible={true}
        focusable={false}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="button-failure"
      >
        <
          color="#ffffff"
          icon={
            Object {
              "icon": Array [
                352,
                512,
                Array [],
                "f00d",
                "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
              ],
              "iconName": "times",
              "prefix": "fas",
            }
          }
          size={100}
        />
      </View>
    </View>
  </View>
  <View />
</View>
`;

exports[`Habbajet Screen Component will return an empty view if no habbajets are selected 1`] = `<View />`;
